File: src/shared/validators/user-organization.validator.ts
import { $Enums } from '@prisma/client';
import { UserRequest } from '../types/user-request';
import { prisma } from '../../infra/database/prisma.service';
import { NotFoundException, UnauthorizedException } from '../exceptions';

export async function userOrganizationValidator(
  organizationId: number,
  userRequest: UserRequest,
) {
  if (!userRequest) throw new UnauthorizedException('Usuário não encontrado!');

  const organization = await prisma.organization.findFirst({
    where: {
      id: organizationId,
      deletedAt: null,
    },
  });

  if (!organization) throw new NotFoundException('Organização não encontrada!');

  if (userRequest.role?.reference === $Enums.RoleReferences.ADMIN) return;

  if (userRequest.organizationId !== organizationId)
    throw new UnauthorizedException('Usuário sem permissão nessa organização');
}

File: src/shared/validators/user_organization_validator.py
from src.modules.organization.models import Organization
from src.modules.role.domain.enums.role_references_enum import RoleReferences
from src.shared.types.user_request import UserRequest
from src.shared.exceptions.unauthorized_exception import UnauthorizedException
from src.shared.exceptions.not_found_exception import NotFoundException


def user_organization_validator(organization_id: int, user_request: UserRequest):

    if user_request is None:
        raise UnauthorizedException("Usuário não encontrado!")

    organization = Organization.objects.filter(
        id=organization_id, deleted_at=None
    ).first()

    if organization is None:
        raise NotFoundException("Organização não encontrada!")

    if user_request.role.reference == RoleReferences.ADMIN:
        return

    if user_request.organization.id != organization_id:
        raise UnauthorizedException("Usuário sem permissão nessa organização")
